### PR TITLE
fix: Memorystore compatibility mode

### PR DESCRIPTION
**Pull request by jasonschroeder-sfdc**
Google Memorystore's `CLUSTER SHARDS` reply is confusing Jedis. Fall back to using `CLUSTER SLOTS` if `CLUSTER SHARDS` throws any `RuntimeException`s.

The bulk of the implementation is resurrected from the implementation pre-https://github.com/bazelbuild/bazel-buildfarm/pull/1677

See also: https://github.com/redis/jedis/issues/3837

### CODE REVIEW

File: src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java [hunk 1]
```diff
@@ -104,7 +111,15 @@ public static List<String> getEvenlyDistributedHashesWithPrefix(
   private static Iterable<List<List<Long>>> getNodeSlotRanges(UnifiedJedis jedis) {
     if (jedis instanceof JedisCluster) {
       // get slot range information for each shard
-      return transform(getClusterShards((JedisCluster) jedis), ClusterShardInfo::getSlots);
+      try {
+        return transform(getClusterShards((JedisCluster) jedis), ClusterShardInfo::getSlots);
+      } catch (RuntimeException re) {
+        log.log(
+            Level.WARNING,
+            "Unable to get Redis Cluster shards. Fall-back to deprecated implementation",
```
Comment (werkt) [inline src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java:119]: nit: "CLUSTER SHARDS failed. Falling back to CLUSTER NODES (deprecated)"

There's enough context in the backtrace to implicate redis.

File: src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java [hunk 2]
```diff
@@ -44,7 +44,7 @@
 @Log
 public class RedisNodeHashes {
   private static final Iterable<List<List<Long>>> SINGLETON_NODE_SLOT_RANGES =
-      ImmutableList.of(ImmutableList.of(ImmutableList.of(0l, CLUSTER_HASHSLOTS - 1l)));
+      Collections.singleton(ImmutableList.of(ImmutableList.of(0l, CLUSTER_HASHSLOTS - 1l)));
```
Comment (werkt) [inline src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java:47]: That rather makes the next interior ImmutableList.of a useful singleton candidate, no?

File: src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java [hunk 3]
```diff
@@ -86,18 +86,18 @@ public static List<String> getEvenlyDistributedHashesWithPrefix(
...
           hashTags.add(RedisSlotToHash.correlateRangesWithPrefix(slotRanges, prefix));
         }
         return hashTags.build();
       } catch (JedisException e) {
-        return ImmutableList.of();
+        return List.of();
       }
     } else {
-      return ImmutableList.of(prefix);
+      return List.of(prefix);
```
Comment (werkt) [inline src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java:100]: I don't agree with this (or really the List response - it should be Iterable) - I don't want the response to be a mutable object.

File: src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java (hunk 2 above)
Comment (jasonschroeder-sfdc) [reply src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java:47]: Curiously, `Collections.singleton(...)` returns a `java.util.Set` of one item! I would not have guessed that.

File: src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java (hunk 3 above)
Comment (jasonschroeder-sfdc) [reply src/main/java/build/buildfarm/common/redis/RedisNodeHashes.java:100]: No worries. IntelliJ keeps recommending this refactor so I thought I'd try it. I'll just fixup this commit.

**Comment by jasonschroeder-sfdc** [comment]
Google Memorystore says they will fix this "soon":
https://github.com/redis/jedis/issues/3837#issuecomment-2130059005

**Comment by jasonschroeder-sfdc** [comment]
Google has fixed Memorystore, this patch is no longer needed.

**Pull request closed without merging**